### PR TITLE
Federated Ingress Controller

### PR DESCRIPTION
**Pull request by ghost**
Based on new federated controller libraries. 

cc @kubernetes/sig-cluster-federation @mfanjie @nikhiljindal @mwielgus @mml @madhusudancs FYI

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30419)

### CODE REVIEW

File: federation/pkg/federation-controller/ingress/ingress-controller.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+	// Build delivereres for triggering reconcilations.
+	ic.ingressDeliverer = util.NewDelayingDeliverer()
+	ic.clusterDeliverer = util.NewDelayingDeliverer()
+
+	// Start informer in federated API servers on ingresss that should be federated.
+	ic.ingressInformerStore, ic.ingressInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").List(options)
```
Comment (mfanjie) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: should use `api.NamespaceAll`?

File: federation/pkg/federation-controller/ingress/ingress-controller.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+	// For triggering all ingresss reconcilation. This is used when
+	// a new cluster becomes available.
+	clusterDeliverer *util.DelayingDeliverer
+
+	// Contains ingresss present in members of federation.
+	ingressFederatedInformer util.FederatedInformer
+	// For updating members of federation.
+	federatedUpdater util.FederatedUpdater
+	// Definitions of ingresss that should be federated.
```
Comment (mfanjie) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: ingress->ingresses

File: federation/pkg/federation-controller/ingress/ingress-controller.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func (ic *IngressController) Stop() {
+	ic.ingressFederatedInformer.Stop()
+	close(ic.stopChan)
+}
+
+func (ic *IngressController) deliverIngressObj(obj interface{}, delay time.Duration, trial int64) {
+	ingress := obj.(*extensions.Ingress)
+	ic.deliverIngress(ingress.Name, delay, trial)
```
Comment (mfanjie) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: should we put namespace+name as the key?

File: federation/pkg/federation-controller/ingress/ingress-controller.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").Watch(options)
+			},
+		},
+		&extensions.Ingress{},
+		controller.NoResyncPeriodFunc(),
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, 0, 0) }))
+
+	// Federated informer on ingresss in members of federation.
+	ic.ingressFederatedInformer = util.NewFederatedInformer(
```
Comment (mfanjie) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: The user agent name of federatedInformer is `userAgentName     = "federation-service-controller"`, we need change this as a more comment name.

File: federation/pkg/federation-controller/ingress/ingress-controller.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: Yes, agreed.  I couldn't remember the correct incantation, so I added that placeholder instead :-)

File: federation/pkg/federation-controller/ingress/ingress-controller.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: Agreed, done.

File: federation/pkg/federation-controller/ingress/ingress-controller.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: Agreed.  Will do.   The controller that this is modelled on is for namespaces themselves, which are not scoped by namespace, hence this.

File: federation/pkg/federation-controller/ingress/ingress-controller.go (hunk 4 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: Agreed, but fix in a separate PR, to keep this PR clean.

File: federation/pkg/federation-controller/ingress/ingress-controller.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").Watch(options)
+			},
+		},
+		&extensions.Ingress{},
+		controller.NoResyncPeriodFunc(),
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, 0, 0) }))
```
Comment (jianhuiz) [inline federation/pkg/federation-controller/ingress/ingress-controller.go]: Do we still need to handle `DeletedFinalStateUnknown` case?

**Comment by quinton-hoole** [comment]
This is ready for review now.  Unit tests pass, although I'm sure we'll find some issues when tested against the new e2e tests in https://github.com/kubernetes/kubernetes/pull/29773.

I will add the following in small followup PR's:

1. DNS record creation
2. namespace handling
3. sharing a public IP across shards

cc @mml @nikhiljindal @madhusudancs

**Comment by quinton-hoole** [comment]
Whoops!  It seems I messed up a branch merge.  Until I fix that up, reviewers need only look at 

1.  federation/pkg/federation-controller/ingress
2.  federation/cmd/federation-controller-manager/app/controllermanager.go

**Comment by quinton-hoole** [comment]
OK, I fixed my merge mess-up.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+	smallDelay            time.Duration
+	updateTimeout         time.Duration
+}
+
+// A structure passed by delaying deliverer. It contains an ingress that should be reconciled and
+// the number of previous attempts that ended up in some kind of ingress-related
+// error (like failure to create).
+type ingressItem struct {
+	ingress      string
+	prevAttempts int64
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: you probably don't need prev attempts (and this struct at all).

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+	if !ic.isSynced() {
+		ic.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ic.clusterAvailableDelay))
+	}
+	for _, obj := range ic.ingressInformerStore.List() {
+		ingress := obj.(*extensions_v1beta1.Ingress)
+		ic.deliverIngress(ingress.Name, ic.smallDelay, false)
+	}
+}
+
+func backoff(trial int64) time.Duration {
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: backoff is now handled with util/lifecycle/backoff. please take a look at namespace controller.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+	if trial > 12 {
+		return 12 * 5 * time.Second
+	}
+	return time.Duration(trial) * 5 * time.Second
+}
+
+func (ic *IngressController) reconcileIngress(ingress string) {
+	glog.V(4).Infof("Reconciling ingress %q", ingress)
+	if !ic.isSynced() {
+		ic.deliverIngress(ingress, ic.clusterAvailableDelay, false)
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: return if not synced

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 7 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 6 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 8 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done.

**Comment by quinton-hoole** [comment]
Thanks for the review @mwielgus .  I've:

1. addressed all of your comments
2. added support for namespaces (as we discussed)
3. fixed a bug in the create and update logic around ObjectMeta (as discussed)
4. added logic to share a single global IP across all shards as described [here](https://gist.github.com/bprashanth/52648b2a0b6a5b637f843e7efb2abc97)

PTAL.  I think this is ready for LGTM now. All unit tests, and manual e2e tests passed. Automated e2e tests are lined up in https://github.com/kubernetes/kubernetes/pull/29773 which should merge in the next couple of days.

**Comment by quinton-hoole** [comment]
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 62 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-62:-KPko-I-DbavhUc1-KtZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L62)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed, done.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 96 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-96:-KPkoSTyq4i153DTX7-l:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L96)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Yes, agreed.  I couldn't remember the correct incantation, so I added that placeholder instead :-)
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 107 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-107:-KPkoWWQg2eD8mG6OL69:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L107)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed, but fix in a separate PR, to keep this PR clean.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 176 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KOrA5OARvhYrPS52sHE-r1-176:-KPkoZyWzAWFSEsslhIx:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/295932a6245e3a59c9b53970ff4610c8461a783b/federation/pkg/federation-controller/ingress/ingress_controller.go#L176)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Agreed.  Will do.   The controller that this is modelled on is for namespaces themselves, which are not scoped by namespace, hence this.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 80 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KPUY_wXsqf5H0npP5Hy-r5-80:-KPkoJjk9jEK7QQVqpU0:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e65ff45a3f89020739036f77e9fe9e0fc0f1af89/federation/pkg/federation-controller/ingress/ingress_controller.go#L80)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Done.
</details>
Done.

---

*[federation/pkg/federation-controller/ingress/ingress_controller.go, line 255 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419#-KPUY_wXsqf5H0npP5Hy-r5-255:-KPkol39L5g90aCggRsn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e65ff45a3f89020739036f77e9fe9e0fc0f1af89/federation/pkg/federation-controller/ingress/ingress_controller.go#L255)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30419)*

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+import (
+	"reflect"
+	"time"
+
+	federation_api "k8s.io/kubernetes/federation/apis/federation/v1beta1"
+	federation_release_1_4 "k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	"k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	// api_v1 "k8s.io/kubernetes/pkg/api/v1"
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: remove

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+				}
+			}
+			// Update existing ingress, if needed.
+			if !util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) ||
+				!reflect.DeepEqual(desiredIngress.Spec, clusterIngress.Spec) {
+				// TODO: In some cases Ingress controllers in the clusters add annotations, so we ideally need to exclude those from
+				// the equivalence comparison to cut down on unnecessary updates.
+				glog.V(4).Infof("Ingress %s in cluster %s needs an update: cluster ingress %v is not equivalent to federated ingress %v", ingress, cluster.Name, clusterIngress, desiredIngress)
+				// We need to use server-created fields from the cluster, not the desired object when updating.
+				desiredIngress.ObjectMeta.ResourceVersion = clusterIngress.ObjectMeta.ResourceVersion
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Do we want to copy this?

**Comment by mwielgus** [comment]
minor comments, LGTM assuming a clean-up PR is submitted shortly.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 9 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yup, done.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 10 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Good catch - I don't know what I was thinking there.  Will fix.

Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Oh no wait, I was correct.  When updating (the cluster ingress), one needs to provide the resource version and UID.

Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Why? In namespace controller i don't do that and it works. Also if you update an object using kubectl you don't need any of these fields.

Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md

The resourceVersion is changed by the server every time an object is modified. If resourceVersion is included with the PUT operation the system will verify that there have not been other successful mutations to the resource during a read/modify/write cycle, by verifying that the current value of resourceVersion matches the specified value.

**Comment by quinton-hoole** [comment]
Fixed the nit, squashed, and checked out the e2e test failure, which looks like a flake - the test cluster never came up.  Applying LGTM as per review feedback above.

**Comment by quinton-hoole** [comment]
Rebased.  Re-applying LGTM.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 10 above)
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: The thing here is that we don't care about the previous content of the object here. We could overwrite the content every time we reconciliate an object but we do this only when something changed for performance reason.

Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes, you're right, I had originally assumed that resourceVersion was mandatory for PUT's, but it's not.  I'll remove this in a followup PR.

**Pull request merged**